tutorials/workflow/csharp/combined-patterns/WorkflowApp/OrderWorkflow.cs

```diff
@@ -0,0 +1,65 @@
+using Dapr.Workflow;
+using WorkflowApp.Activities;
+
+namespace WorkflowApp;
+
+internal sealed class OrderWorkflow : Workflow<Order, OrderStatus>
+{
+   public override async Task<OrderStatus> RunAsync(WorkflowContext context, Order order)
+   {
+       var inventoryTask = context.CallActivityAsync<ActivityResult>(
+               nameof(CheckInventory),
+               order.OrderItem);
+        var shippingDestinationTask = context.CallActivityAsync<ActivityResult>(
+               nameof(CheckShippingDestination),
+               order);
+       List<Task<ActivityResult>> tasks = [inventoryTask, shippingDestinationTask];
+       var taskResult = await Task.WhenAll(tasks);
+
+       if (taskResult.Any(r => !r.IsSuccess))
+       {
+           var message = $"Order processing failed. Reason: {taskResult.First(r => !r.IsSuccess).Message}";
+           return new OrderStatus(IsSuccess: false, message);
+       }
+
+       var paymentResult = await context.CallActivityAsync<PaymentResult>(
+           nameof(ProcessPayment),
+           order);
+       if (paymentResult.IsSuccess)
+       {
+           var inventoryResult = await context.CallActivityAsync<UpdateInventoryResult>(
+               nameof(UpdateInventory),
+               order.OrderItem);
+       }
+
+       ShipmentRegistrationStatus shipmentRegistrationStatus;
+       try
+       {
+           await context.CallActivityAsync<RegisterShipmentResult>(
+               nameof(RegisterShipment),
+               order);
+           shipmentRegistrationStatus = await context.WaitForExternalEventAsync<ShipmentRegistrationStatus>(
+               eventName: Constants.SHIPMENT_REGISTERED_EVENT,
+               timeout: TimeSpan.FromSeconds(300));
+       }
+       catch (TaskCanceledException)
+       {
+           // Timeout occurred
+           var message = $"ShipmentRegistrationStatus for {order.Id} timed out.";
```

Comment:
This should throw instead of returning a type so that the workflow will re-run and try again.

Reply:
This exception is due to the `WaitForExternalEvent` timing out when no event has been received. So we don't want to re-run the workflow again.